File: python/sorting_algorithms/non-comparative_sorting_algorithms/bucketSort.py
def bucket_sort(arr):
    n = len(arr)
    if n <= 0:
        return arr

    # Crear cubetas vacias
    num_buckets = 10
    buckets = [[] for _ in range(num_buckets)]

    # Distribuir elementos en cubetas
    for num in arr:
        index = int(num * num_buckets)
        buckets[index].append(num)

    # Ordenar cada cubeta y concatenar resultados
    sorted_arr = []
    for bucket in buckets:
        sorted_arr.extend(sorted(bucket))

    return sorted_arr

# Ejemplo de uso
arr = [0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68]
print("\nArray original:", arr)
sorted_arr = bucket_sort(arr)
print("\nArray ordenado:", sorted_arr)

File: c/sorting_algorithms/non-comparative_sorting_algorithms/bucketSort.c
#include <stdio.h>
#include <stdlib.h>

// Funcion para ordenar una cubeta usando Insertion Sort
void insertionSort(float arr[], int n) {
    for (int i = 1; i < n; i++) {
        float key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// Bucket Sort
void bucketSort(float arr[], int n) {
    // Crear cubetas
    int numBuckets = 10;
    float* buckets[numBuckets];
    int bucketSizes[numBuckets];

    for (int i = 0; i < numBuckets; i++) {
        buckets[i] = (float*)malloc(n * sizeof(float));
        bucketSizes[i] = 0;
    }

    // Distribuir elementos en cubetas
    for (int i = 0; i < n; i++) {
        int bucketIndex = (int)(arr[i] * numBuckets);
        buckets[bucketIndex][bucketSizes[bucketIndex]++] = arr[i];
    }

    // Ordenar cubetas y concatenar resultados
    int index = 0;
    for (int i = 0; i < numBuckets; i++) {
        insertionSort(buckets[i], bucketSizes[i]);
        for (int j = 0; j < bucketSizes[i]; j++) {
            arr[index++] = buckets[i][j];
        }
        free(buckets[i]);
    }
}

// Imprimir un array
void printArray(float arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%.2f ", arr[i]);
    }
    printf("\n");
}

int main() {
    float arr[] = {0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("\nArray original:\n");
    printArray(arr, n);

    bucketSort(arr, n);

    printf("\nArray ordenado:\n");
    printArray(arr, n);

    return 0;
}